File: src/class_gauss_jacobi.f90
!=====================================================================!
! Diagonal-sweep linear solver (traditionally: Gauss-Jacobi): supplies
! only the sweep (`iterate`) - inverting the self-loop subgraph of the
! operator - and inherits the residual-minimization iteration from
! linear_solver.
!=====================================================================!

module class_gauss_jacobi

  use iso_fortran_env         , only : dp => REAL64
  use interface_linear_solver , only : linear_solver
  use interface_assembler     , only : assembler, DIAGONAL, LOWER_TRIANGLE, UPPER_TRIANGLE

  implicit none

  ! Expose only the linear solver datatype
  private
  public :: gauss_jacobi

  !===================================================================!
  ! Linear solver datatype
  !===================================================================!

  type, extends(linear_solver) :: gauss_jacobi

   contains

     ! the sweep consumed by the inherited outer iteration
     procedure :: iterate

  end type gauss_jacobi

  !===================================================================!
  ! Interface for multiple constructors
  !===================================================================!

  interface gauss_jacobi
     module procedure construct
  end interface gauss_jacobi

contains

  !===================================================================!
  ! Constructor for linear solver
  !===================================================================!

  pure type(gauss_jacobi) function construct(max_it, &
       & max_tol, print_level) result (this)

    type(integer)  , intent(in) :: max_it
    type(real(dp)) , intent(in) :: max_tol
    type(integer)  , intent(in) :: print_level

    this % max_it      = max_it
    this % max_tol     = max_tol
    this % print_level = print_level
    this % res_file    = 'gj.res'

  end function construct

  !===================================================================!
  ! The sweep: jacobi iteration on the correction equation A dx = r
  ! from dx = 0, inverting the diagonal (the self-loops) each pass.
  !===================================================================!

  impure subroutine iterate(this, system, r, dx, iter)

    class(gauss_jacobi) , intent(inout) :: this
    class(assembler)    , intent(in)  :: system
    real(dp)            , intent(in)  :: r(:)
    real(dp)            , intent(out) :: dx(:)
    integer             , intent(out) :: iter

    ! Locals
    real(dp) :: tol, bnorm
    real(dp) , allocatable :: Ux(:), Lx(:), D(:), R2(:), xnew(:), identity(:)

    dx = 0.0_dp
    allocate(Ux, Lx, D, R2, xnew, identity, mold = dx)

    ! Extract the diagonal entries (the self-loop subgraph on ones)
    identity = 1.0d0
    call system % get_jacobian_residual_product(D, identity, part = DIAGONAL)

    bnorm = sqrt(system % inner_product(r, r))

    ! Homogeneous case (nothing to do)
    if (bnorm .le. this % max_tol) then
       iter = 0
       return
    end if

    !-----------------------------------------------------------------!
    ! Apply the jacobi sweep until tolerance is achieved
    !-----------------------------------------------------------------!

    iter = 1; tol  = huge(1.0d0)
    do while ((tol .gt. this % max_tol) .and. (iter .lt. this % max_it))

       ! Form the residual (partial) after the split
       call system % get_jacobian_residual_product(Ux, dx, part = UPPER_TRIANGLE)
       call system % get_jacobian_residual_product(Lx, dx, part = LOWER_TRIANGLE)

       R2 = r - Lx - Ux

       ! Invert the diagonal
       xnew = R2/D ! D^{-1}(r - L dx - U dx)
       tol  = sqrt(system % inner_product(dx - xnew, dx - xnew))

       if (this % print_level .gt. 1) then
          write(*,*) "inner (1)", iter, tol
       end if

       dx   = xnew
       iter = iter + 1

    end do

    deallocate(Ux, Lx, D, R2, xnew, identity)

  end subroutine iterate

end module class_gauss_jacobi

File: ufvm/class_gauss_jacobi.py
#=====================================================================#
# Gauss-Jacobi linear solver class that uses the functionalities of
# assembler class in the iterative solution process.
#=====================================================================#

import numpy as np

from .interface_linear_solver import LinearSolver


class GaussJacobi(LinearSolver):

    #===================================================================!
    # Constructor for linear solver
    #===================================================================!
    def __init__(self, FVAssembler, max_it, max_tol, print_level):
        self.FVAssembler = FVAssembler
        self.max_it = max_it
        self.max_tol = max_tol
        self.print_level = print_level

    #===================================================================!
    # Outer (skew-source) iterations
    #===================================================================!
    def solve(self):
        # Initial guess vector for the subspace is "b"
        x = np.zeros(self.FVAssembler.num_state_vars, dtype=np.float64)
        self.FVAssembler.get_source(x)
        if np.linalg.norm(x) < np.finfo(np.float64).eps:
            print('zero rhs? stopping')
            raise SystemExit

        xold = np.zeros(self.FVAssembler.num_state_vars, dtype=np.float64)
        ss = np.zeros(self.FVAssembler.num_state_vars, dtype=np.float64)

        res_file = None
        if self.print_level == -1:
            res_file = open('gj.res', 'w')
            res_file.write(" iteration  residual\n")

        update_norm = np.finfo(np.float64).max
        iter = 1
        while update_norm > self.max_tol:

            xold[:] = x

            if iter == 1:
                ss[:] = 0.0
                num_inner_iters = self.iterate(x, ss)
            else:
                self.FVAssembler.get_skew_source(ss, x)
                num_inner_iters = self.iterate(x, ss)

            update_norm = np.linalg.norm(x - xold)
            if self.print_level == -1:
                res_file.write("%d %s\n" % (iter, repr(float(update_norm))))
            if self.print_level > 0:
                print("outer iter", iter, "num_inner_iters", num_inner_iters,
                      "update norm", update_norm)
            iter = iter + 1

        if self.print_level == -1:
            res_file.close()

        return x

    #===================================================================!
    # Gauss-Jacobi inner iteration
    #===================================================================!
    def iterate(self, x, ss):
        b = np.zeros_like(x)
        Ux = np.zeros_like(x)
        Lx = np.zeros_like(x)
        D = np.zeros_like(x)
        identity = np.zeros_like(x)

        # Identity vector
        identity[:] = 1.0

        # Extract the diagonal entries
        self.FVAssembler.get_jacobian_vector_product(
            D, identity, filter=self.FVAssembler.DIAGONAL)

        # Assemble RHS of the linear system (source + boundary terms)
        self.FVAssembler.get_source(b)

        # Add the skew source terms if supplied
        b = b + ss

        bnorm = np.linalg.norm(b)

        # Homogeneous case (nothing to do)
        if bnorm <= self.max_tol:
            x[:] = 0.0
            return 1

        #-----------------------------------------------------------------!
        # Apply Gauss Jacobi Iterative scheme until tolerance is achieved
        #-----------------------------------------------------------------!
        iter = 1
        tol = np.finfo(np.float64).max
        while (tol > self.max_tol) and (iter < self.max_it):

            # Form the residual (partial) after split
            self.FVAssembler.get_jacobian_vector_product(
                Ux, x, filter=self.FVAssembler.UPPER_TRIANGLE)

            self.FVAssembler.get_jacobian_vector_product(
                Lx, x, filter=self.FVAssembler.LOWER_TRIANGLE)

            R = b - Lx - Ux

            # Invert diagonal: D^{-1}(b - Lx - Ux)
            xnew = R / D
            tol = np.linalg.norm(x - xnew)

            if self.print_level > 1:
                print("inner (1)", iter, tol)

            x[:] = xnew
            iter = iter + 1

        return iter
